interfaces/prompting/patterns_test.go

```diff
@@ -0,0 +1,1365 @@
+				"/foo/bar*/**/baz/",
+			},
+			"/foo/bar/**/",
+		},
+		{
+			[]string{
+				"/foo/bar/**",
+				"/foo/bar*/**/baz/",
+			},
+			"/foo/bar/**",
+		},
+		// Globs followed by doublestars
+		{
+			[]string{
+				"/foo/bar*/**/baz",
+				"/foo/bar*/**/",
+			},
+			"/foo/bar*/**/baz",
+		},
+		{
+			[]string{
+				"/foo/bar*/**/",
+				"/foo/bar*/**",
+			},
+			"/foo/bar*/**/",
+		},
+		// Miscellaneous
+		{
+			[]string{
+				"/foo/bar/*.gz",
+				"/foo/bar/*.tar.gz",
+			},
+			"/foo/bar/*.tar.gz",
+		},
+		{
+			[]string{
+				"/foo/bar/**/*.gz",
+				"/foo/**/*.tar.gz",
+			},
+			"/foo/bar/**/*.gz",
+		},
+		{
+			[]string{
+				"/foo/bar/x/**/*.gz",
+				"/foo/bar/**/*.tar.gz",
+			},
+			"/foo/bar/x/**/*.gz",
+		},
+		{
+			// Both match `/foo/bar/baz.tar.gz`
+			[]string{
+				"/foo/bar/**/*.tar.gz",
+				"/foo/bar/*",
+			},
+			"/foo/bar/*",
+		},
+		{
+			[]string{
+				"/foo/bar/**",
+				"/foo/bar/baz/**",
+				"/foo/bar/baz/**/*.txt",
+			},
+			"/foo/bar/baz/**/*.txt",
+		},
+		{
+			// both match /foo/bar
+			[]string{
+				"/foo/bar*",
+				"/foo/bar/**",
+			},
+			"/foo/bar*",
+		},
+		{
+			[]string{
+				"/foo/bar/*/baz*/**/fizz/*buzz",
+				"/foo/bar/*/baz*/**/fizz/bu*zz",
+				"/foo/bar/*/baz*/**/fizz/buzz",
+				"/foo/bar/*/baz*/**/fizz/buzz*",
+			},
+			"/foo/bar/*/baz*/**/fizz/buzz",
+		},
+		{
+			[]string{
+				"/foo/*/bar/**",
+				"/foo/**/bar/*",
+			},
+			"/foo/*/bar/**",
+		},
+		{
+			[]string{
+				`/foo/\\\b\a\r`,
+				`/foo/barbaz`,
+			},
+			`/foo/barbaz`,
+		},
+		{
+			[]string{
+				`/foo/\\`,
+				`/foo/*/bar/x`,
+			},
+			`/foo/\\`,
+		},
+		{
+			[]string{
+				`/foo/\**/b\ar/*\*`,
+				`/foo/*/bar/x`,
+			},
+			`/foo/\**/b\ar/*\*`,
+		},
+		// Patterns with "**[^/]" should not be emitted from ExpandPathPattern
+		{
+			[]string{
+				"/foo/**",
+				"/foo/**bar",
+			},
+			"/foo/**bar",
+		},
+		// Duplicate patterns should never be passed into GetHighestPrecedencePattern,
+		// but if they are, handle them correctly.
+		{
+			[]string{
+				"/foo/bar/",
+				"/foo/bar/",
+			},
+			"/foo/bar/",
+		},
+		{
+			[]string{
+				"/foo/bar/",
+				"/foo/bar/",
+				"/foo/bar",
+			},
+			"/foo/bar/",
+		},
+		{
+			[]string{
+				"/foo/bar/**",
+				"/foo/bar/**",
+				"/foo/bar/*",
+			},
+			"/foo/bar/*",
+		},
+	} {
+		highestPrecedence, err := prompting.GetHighestPrecedencePattern(testCase.patterns)
+		c.Check(err, IsNil, Commentf("Error occurred during test case %d:\n%+v", i, testCase))
+		c.Check(highestPrecedence, Equals, testCase.highestPrecedence, Commentf("Highest precedence pattern incorrect for test case %d:\n%+v", i, testCase))
+	}
+
+	orderedPatterns := []string{
+		"/foo/bar/baz/myfile.txt",
+		"/foo/bar/baz/m?file.*",
+		"/foo/bar/baz/m*file.txt",
+		"/foo/bar/baz/m*file*",
+		"/foo/bar/baz/*",
+		"/foo/bar/*/myfile.txt",
+		"/foo/bar/*/myfile*",
+		"/foo/bar/*/*",
+		"/foo/bar/**/baz/myfile.txt",
+		"/foo/bar/**/myfile.txt",
+		"/foo/bar/**/baz/*.txt",
+		"/foo/bar/**/baz/*",
```

Comment:
shouldn't `"/foo/bar/**/myfile.txt"` follow `"/foo/bar/**/baz/*"` and precede `"/foo/bar/**"`. TBH I'm not even sure what is the proper outcome of precedence of rules in say line 843 and 845.

Reply:
Yes, I agree, since `/foo/bar/**/baz/*` matches `baz` earlier in the path than `myfile.txt`, it should have precedence. So line 845 has precedence over line 843. This should be correct now since the reimplementation.